File: helfomat-infrastructure-filesystem/src/test/java/de/helfenkannjeder/helfomat/infrastructure/filesystem/RestDownloadServiceTest.java
package de.helfenkannjeder.helfomat.infrastructure.filesystem;

import de.helfenkannjeder.helfomat.api.geopoint.DistanceMatrixApplicationService;
import de.helfenkannjeder.helfomat.api.picture.RestDownloadService;
import org.assertj.core.api.ThrowableAssert;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.web.client.HttpClientErrorException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * @author Valentin Zickner
 */
@SpringBootTest
@ExtendWith(SpringExtension.class)
class RestDownloadServiceTest {

    @Autowired
    private RestDownloadService downloadService;

    @MockBean
    private DistanceMatrixApplicationService distanceMatrixApplicationService;

    @BeforeAll
    static void setUpServer() throws Exception {
        EmbeddedHttpServer.start();
        EmbeddedHttpServer.setContent("/Radlader_2124_5x2.jpg", null, "de/helfenkannjeder/picture/Radlader_2124_5x2.jpg");
    }

    @AfterAll
    static void tearDownServer() {
        EmbeddedHttpServer.stop();
    }

    @Test
    void downloadPicture_withCorrectUrl_returnsByteArray() {
        // Act
        byte[] bytes = downloadService.download(getUrl("Radlader_2124_5x2.jpg"));

        // Assert
        assertThat(bytes).isNotNull();
        assertThat(bytes.length).isEqualTo(2838526);
        assertThat(bytes[0]).isEqualTo((byte) 0xFF);
        assertThat(bytes[1]).isEqualTo((byte) 0xD8);
    }

    @Test
    void downloadPicture_withIncorrectUrl_throwsHttpClientErrorException() {
        // Act
        ThrowableAssert.ThrowingCallable throwable = () ->
            downloadService.download(getUrl("test.jpg"));

        // Assert
        assertThatThrownBy(throwable).isInstanceOf(HttpClientErrorException.class);
    }

    private static String getUrl(String url) {
        return "http://localhost:" + EmbeddedHttpServer.PORT + "/" + url;
    }
}

File: helfomat-infrastructure-filesystem/src/test/kotlin/de/helfenkannjeder/helfomat/infrastructure/filesystem/RestDownloadServiceTest.kt
package de.helfenkannjeder.helfomat.infrastructure.filesystem

import de.helfenkannjeder.helfomat.api.picture.RestDownloadService
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.ThrowableAssert.ThrowingCallable
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.web.client.HttpClientErrorException

/**
 * @author Valentin Zickner
 */
@SpringBootTest
@ExtendWith(SpringExtension::class)
internal class RestDownloadServiceTest {

    @Autowired
    private lateinit var downloadService: RestDownloadService


    @Test
    fun downloadPicture_withCorrectUrl_returnsByteArray() {
        // Act
        val bytes: ByteArray? = downloadService.download(getUrl("Radlader_2124_5x2.jpg"))

        // Assert
        assertThat(bytes?.size).isEqualTo(2838526)
        assertThat(bytes?.get(0)).isEqualTo(0xFF.toByte())
        assertThat(bytes?.get(1)).isEqualTo(0xD8.toByte())
    }

    @Test
    fun downloadPicture_withIncorrectUrl_throwsHttpClientErrorException() {
        // Act
        val throwable = ThrowingCallable { downloadService.download(getUrl("test.jpg")) }

        // Assert
        Assertions.assertThatThrownBy(throwable).isInstanceOf(HttpClientErrorException::class.java)
    }

    companion object {

        private val EMBEDDED_HTTP_SERVER = EmbeddedHttpServer()

        @BeforeAll
        @JvmStatic
        fun setUpServer() {
            EMBEDDED_HTTP_SERVER.start()
            EMBEDDED_HTTP_SERVER.setContent("/Radlader_2124_5x2.jpg", null, "de/helfenkannjeder/picture/Radlader_2124_5x2.jpg")
        }

        @AfterAll
        @JvmStatic
        fun tearDownServer() {
            EMBEDDED_HTTP_SERVER.stop()
        }

        private fun getUrl(url: String): String {
            return "http://localhost:${EMBEDDED_HTTP_SERVER.port}/$url"
        }

    }
}